{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "outside-dispatch",
   "metadata": {},
   "source": [
    "# Specification\n",
    "\n",
    "- This is a command Line program using standard parameters, STDIN and STDOUT redirection\n",
    "- Program name: missingMotif.py\n",
    "- Input/Output: missingMotif.py < [input file] > [output file]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "humanitarian-think",
   "metadata": {},
   "source": [
    "- --minMotif  minimum motif size to evaluate (int)\n",
    "- --maxMotif maximum motif size to evaluate (int)\n",
    "- --cutoff  Z-score cutoff (float)\n",
    "\n",
    "Ex: missingMotif.py --minMotif=3 --maxMotif=8 --cutoff=-4.0 < [input file] > [output file]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "reliable-directive",
   "metadata": {},
   "source": [
    "# Background\n",
    "As cutting edge genomic sequencing technologies mature, their accessibility is enhanced while costs shrink, in contrast to specialized technologies such as Single-molecule real-time (SMRT) that reveal methylomes. The application of methylomic data to predict R-M sites is established, however, costly overheads prevent adoption on the same scale as genomic sequencing technologies.\n",
    "\n",
    "In contrast to established methods for R-M components,Stealth performs a statistical comparison between the genomic counts of motifs within a specified size range (k-mers), and probabilities based on a genome specific null model derived from a markov (k-2) chain. Stealth outputs a list of motifs that are systematically underrepresented in the genome, which are expected to include novel R-M substrate motifs. In a 2022 paper, Bernick showed that synthetic DNA designed to avoid some of these underrepresented motifs in H. Pylori yielded improvements in transformation efficiency of >10,000%.\n",
    "\n",
    "# The signifigance of under-represented motifs\n",
    "\n",
    "RM systems are typically composed of a suite of restriction nucleases and cognate methyltransferases. The restriction component of these RM systems provide a defense strategy by selectively cleaving DNA at recognition sites, often characterized by short 4-6 base pair reverse complement palindromic motifs. On the other hand, the modification component shields the host genome by enzymatically modifying potential cleavage sites within the host DNA. Meselson and Stahl (1958) demonstrated the semi-conservative model of bacterial double-stranded DNA genome replication [3] . This process yields chromosomal copies that are transiently hemi-methylated, as the nascent strand of the daughter chromosomes have not yet had the opportunity to undergo protective methylation by the RM system. What ensues is a race against time—a race between the RM cleavage machinery and the protective DNA methylation processes. This molecular interplay imparts a selective pressure on the genome towards reduction of endogenous RM recognition motifs, leading to the underrepresentation of RM motifs within the genome over time, allowing them to be identified as systemically underrepresented in genomic sequence.\n",
    "\n",
    "Note: \"Stealth\" has recently been accepted for publication, and is now publicly available at: https://git.ucsc.edu/dbernick/stealth\n",
    "\n",
    "# The following is my implimentation of a HMM based approach to identify such motifs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "absolute-prompt",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Example output\n",
    "sequence: reverse   count   Expect  Zscore\n",
    "AAAATTTA:TAAATTTT\t835\t    1737.66\t-21.66\n",
    "GATTAATA:TATTAATC\t550\t    1326.89\t-21.33\n",
    "AATTAATA:TATTAATT\t929\t    1839.72\t-21.24\n",
    "AATTAATC:GATTAATT\t977\t    1861.79\t-20.51\n",
    "ATAATTAA:TTAATTAT\t1033    1926.49\t-20.36\n",
    "GAAATTTA:TAAATTTC\t378 \t1031.07\t-20.34\n",
    "CCGATCGC:GCGATCGG\t1323\t2284.74\t-20.12\n",
    "GCGATCGA:TCGATCGC\t1221\t2121.78\t-19.56\n",
    "ACGATCGC:GCGATCGT\t1188\t2035.31\t-18.78\n",
    "GTTTAAAA:TTTTAAAC\t519\t    1151.17\t-18.63\n",
    "AAATATTA:TAATATTT\t738\t    1444.79\t-18.60\n",
    "AAAATTTG:CAAATTTT\t857\t    1591.95\t-18.42\n",
    "CATTAATC:GATTAATG\t479\t    1037.49\t-17.34\n",
    "ATATATAA:TTATATAT\t271\t    736.19\t-17.15\n",
    "ATTTAAAG:CTTTAAAT\t419\t    941.16\t-17.02\n",
    "AAATATTG:CAATATTT\t687\t    1287.74\t-16.74\n",
    "CAAATTTA:TAAATTTG\t670\t    1265.24\t-16.74\n",
    "CTTTAAAC:GTTTAAAG\t468\t    995.13\t-16.71"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "medieval-running",
   "metadata": {},
   "source": [
    "This cell is adapted from Dr. Bernicks derivation provided in class\n",
    "# Null Model derivation\n",
    "If we assume complete independence of the any of the characters that make up a sequence K, we could approximate the expected count of K, E(K),  as:\n",
    "$\\begin{matrix} \n",
    "K=k_{1}k_{2}k_{3}k_{4}, Pr(k_{i})=\\frac{c(k_{i})}{N} \\\\ \n",
    "E(K)=N * Pr(K) = N * \n",
    "Pr(k_{1})\n",
    "Pr(k_{2})\n",
    "Pr(k_{3})\n",
    "Pr(k_{4}) \n",
    "\\end{matrix} \n",
    "$\n",
    "The previous statement outlines a Markov(0) approximation for the expected count (E) of our sequence K. \n",
    "\n",
    "This assumption of independence is often inadequate due to factors such as codon bias influencing the selection of 3-mers based on underlying amino acid frequencies. To address this, we employ a Markov model, where the probability of any symbol is dependent on the preceding characters. For a k-mer of size 4, we condition our character probabilities using the preceding two bases, obviating the need for the count of the specific 4-mer. Edge conditions are essential to consider in our approximation. Generally, we can utilize a model of order k-2, where k is the size of the k-mer, as long as we possess sufficient data. \n",
    "\n",
    "$\n",
    "\\begin{matrix} \n",
    "Pr(K)=\\frac{E(K)}{N} = \n",
    "Pr(k_{1})\n",
    "Pr(k_{2}|k_{1})\n",
    "Pr(k_{3}|k_{1}k_{2})\n",
    "Pr(k_{4}|k_{2}k_{3}) \\\\\n",
    "= \\frac{c(k_{1})}{N}\n",
    "\\frac{c(k_{1}k_{2})}{c(k_{1})}\n",
    "\\frac{c(k_{1} k_{2} k_{3})} {c(k_{1} k_{2})}\n",
    "\\frac{c(k_{2} k_{3} k_{4})} {c(k_{2} k_{3})} \\\\\n",
    "= \\frac{1}{N}\n",
    "\\frac{c(k_{1} k_{2} k_{3})c(k_{2} k_{3} k_{4})} {c(k_{2} k_{3})}\n",
    "\\end{matrix}\n",
    "$\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aadb6594",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Created on Wed Oct  4 17:22:02 2023\n",
    "@author: jlarbale\n",
    "\"\"\"\n",
    "\n",
    "import sys\n",
    "from itertools import product\n",
    "import math\n",
    "\n",
    "# FastAreader by Dr. David Bernick\n",
    "class FastAreader:\n",
    "    def __init__(self, fname=''):\n",
    "        '''contructor: saves attribute fname '''\n",
    "\n",
    "        self.fname = fname\n",
    "        self.fileH = None\n",
    "\n",
    "    def doOpen(self):\n",
    "        if self.fname == '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "\n",
    "    def readFasta(self):\n",
    "\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "\n",
    "        with self.doOpen() as self.fileH:\n",
    "\n",
    "            header = ''\n",
    "            sequence = ''\n",
    "\n",
    "            # skip to first fasta header\n",
    "            line = self.fileH.readline()\n",
    "            while not line.startswith('>'):\n",
    "                line = self.fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            for line in self.fileH:\n",
    "                if line.startswith('>'):\n",
    "                    yield header, sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else:\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "\n",
    "        yield header, sequence\n",
    "\n",
    "\n",
    "class CommandLine():\n",
    "    '''\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse, now standard in 2.7 and beyond. \n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help, and an error termination mechanism do-usage_and_die.\n",
    "\n",
    "    attributes:\n",
    "    all arguments received from the commandline using .add_argument will be\n",
    "    avalable within the .args attribute of object instantiated from CommandLine.\n",
    "    For example, if myCommandLine is an object of the class, and requiredbool was\n",
    "    set as an option using add_argument, then myCommandLine.args.requiredbool will\n",
    "    name that option.\n",
    "\n",
    "    '''\n",
    "\n",
    "    def __init__(self, inOpts=None):\n",
    "        '''\n",
    "        CommandLine constructor.\n",
    "        Implements a parser to interpret the command line argv string using argparse.\n",
    "        '''\n",
    "\n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(\n",
    "            description='Program prolog - a brief description of what this thing does',\n",
    "            epilog='Program epilog - some other stuff you feel compelled to say',\n",
    "            add_help=True,  # default is True\n",
    "            prefix_chars='-',\n",
    "            usage='%(prog)s [options] -option1[default] <input >output'\n",
    "        )\n",
    "\n",
    "        self.parser.add_argument('-l', '--minMotif', nargs='?', default=1, action='store',\n",
    "                                 help='min kMer size ')\n",
    "        self.parser.add_argument('-m', '--maxMotif', nargs='?', default=8, action='store',\n",
    "                                 help='max kMer size ')\n",
    "        self.parser.add_argument('-c', '--cutoff', nargs='?', type=float, default=.01, action='store',\n",
    "                                 help='Zscore cutoff')\n",
    "\n",
    "        self.parser.add_argument(\n",
    "            '-v', '--version', action='version', version='%(prog)s 0.1')\n",
    "        if inOpts is None:\n",
    "            self.args = self.parser.parse_args()\n",
    "        else:\n",
    "            self.args = self.parser.parse_args(inOpts)\n",
    "\n",
    "\n",
    "def revcomp(kmer):\n",
    "    transDic = {'A': 'T',\n",
    "                'T': 'A', \n",
    "                'G': 'C', \n",
    "                'C': 'G'}\n",
    "    return \"\".join([transDic[i] for i in kmer[::-1]])\n",
    "\n",
    "\n",
    "def key(kmer):\n",
    "    exp = [kmer, revcomp(kmer)] #expand,\n",
    "    return sorted(exp)[0]       #collapse for alphabetical leader\n",
    "\n",
    "\n",
    "class Genome:\n",
    "    '''\n",
    "    reads a fasta file and ranks motifs based on how statistically underrepresented the specific motif is. \n",
    "    We need only consider sequence motifs that are up to about 8 in length ( this is a program option). In \n",
    "    order to use the equation that we developed in class, the minimum size should be at least 3 ( this is an \n",
    "    option also). We will find that there are quite a few of these, so we need to specify a statistical cutoff. \n",
    "    For this assignment, we will use Z-scores. The scores we are looking for will be negative since we are \n",
    "    looking for underrepresented sequences (this cutoff is a program option too). Finally, we need to sort \n",
    "    the output by decreasing motif size and then increasing (most negative to least negative) z-score within \n",
    "    motif size groups and print it out.\n",
    "    '''\n",
    "    def __init__(self, kMin, kMax, psudo=0):\n",
    "        self.psudo = psudo\n",
    "        self.kMin = int(kMin)\n",
    "        self.kMax = int(kMax)\n",
    "        self.kMers = []  # reference kmers\n",
    "        self.kCounts = {}\n",
    "        self.genomicSum = 0\n",
    "\n",
    "\n",
    "        for k in range(self.kMin-2, self.kMax+1):  # inclusivity?\n",
    "            kmerCombinations = product('ACGT', repeat=k) # Generate all possible k-mers of length k using 'ACGT'\n",
    "            self.kMers.extend([''.join(kmer) for kmer in kmerCombinations])\n",
    "\n",
    "        for kMer in self.kMers:\n",
    "            countKey = key(kMer)\n",
    "            self.kCounts[countKey] = self.psudo\n",
    "\n",
    "\n",
    "    def countKmers(self, sequence):\n",
    "        ''' \n",
    "        Recieves genome 1 record at a time. Populate class attribute countsK with counts of each k-mer.\n",
    "        '''\n",
    "        self.genomicSum += len(sequence) \n",
    "        \n",
    "        for i in range(len(sequence)):\n",
    "            for l in range(self.kMin-2, self.kMax+1):\n",
    "                kMer = sequence[i: i + l]\n",
    "                countKey = key(kMer)\n",
    "                self.kCounts[countKey] += 1\n",
    "\n",
    "\n",
    "    def calcE(self):\n",
    "        '''\n",
    "        calculate expected counts based on Markov(k-2) derived relationship\n",
    "        '''\n",
    "        results = []\n",
    "        ref = []\n",
    "\n",
    "        for kMer in self.kMers:\n",
    "            if len(kMer)>self.kMin-1:\n",
    "                try:\n",
    "                    headCount = self.kCounts[key(kMer[:-1])]\n",
    "                    tailCount = self.kCounts[key(kMer[1:])]\n",
    "                    bodyCount = self.kCounts[key(kMer[1:-1])]\n",
    "                    e = (headCount * tailCount) / bodyCount  # mu\n",
    "                except ZeroDivisionError:\n",
    "                    e = 0\n",
    "\n",
    "                stddev = math.sqrt((e*1-e/self.genomicSum))\n",
    "                actual = self.kCounts[key(kMer)]\n",
    "                zscore = (actual-e)/(stddev) if stddev > 0 else 0\n",
    "                \n",
    "                if key(kMer) not in ref:\n",
    "                    results.append([key(kMer), actual, e, zscore])\n",
    "                ref.append(key(kMer))\n",
    "\n",
    "        sortedResults = sorted(results, key=lambda x: (len(x[0]), -x[3], x[0]))\n",
    "        return sortedResults[::-1]\n",
    "\n",
    "import argparse\n",
    "def main():\n",
    "    parser = argparse.ArgumentParser(description=\"Find underrepresented motifs in a genome.\")\n",
    "    parser.add_argument(\"--minMotif\", type=int, required=True, help=\"Minimum motif size\")\n",
    "    parser.add_argument(\"--maxMotif\", type=int, required=True, help=\"Maximum motif size\")\n",
    "    parser.add_argument(\"--cutoff\", type=float, required=True, help=\"Cutoff value\")\n",
    "    parser.add_argument(\"--input\", type=str, required=True, help=\"Input file\")\n",
    "\n",
    "   \n",
    "    args = parser.parse_args()\n",
    "\n",
    "\n",
    "    thisGenome = Genome(args.minMotif, args.maxMotif)\n",
    "\n",
    "    fasta_reader = FastAreader(args.input)\n",
    "    for head, seq in fasta_reader.readFasta():\n",
    "        thisGenome.countKmers(seq)\n",
    "\n",
    "    result = thisGenome.calcE()\n",
    "\n",
    "    \n",
    "    for item in result:\n",
    "        pVal = item[3]\n",
    "        if pVal <= args.cutoff:\n",
    "            seq = item[0]\n",
    "            rSeq = revcomp(item[0])\n",
    "            count = item[1]\n",
    "            E = item[2]\n",
    "            print('{0:8}:{1:8}\\t{2:0d}\\t{3:0.2f}\\t{4:0.2f}'.format(seq, rSeq, count, E, pVal))\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a575c870",
   "metadata": {},
   "outputs": [],
   "source": [
    "Ex: missingMotif.py --minMotif=3 --maxMotif=8 --cutoff=-4.0 < [input file] > [output file]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
